File: 4/solution.jl
function fully_contains(x::Set, y::Set)::Bool
    ⊆(x, y) || ⊇(x, y)
end

section_sets = map(
    line -> map(
        x -> Set(range(map(y -> parse(Int64, y), split(x, "-"))...)),
        split(line, ",")
    ),
    readlines(ARGS[1])
)

# part 1
println(count(x -> fully_contains(x...), section_sets))

# part 2
println(count(v -> !isempty(intersect(v...)), section_sets))

File: 4/solution.py
import sys

def range_set(str_range):
    s1, s2 = str_range.split("-")
    return set(range(int(s1), int(s2) + 1))


section_sets = [
    tuple(map(range_set, line.split(",")))
    for line in open(sys.argv[1], "r")
]

# part 1
print(sum(a <= b or b <= a for a, b in section_sets))

# part 2
print(sum(bool(a & b) for a, b in section_sets))

